### PR TITLE
Upgrade to .NET Standard

### PR DESCRIPTION
**Pull request by StefH**
Upgrade to .NET Standard

### CODE REVIEW

File: vtortola.WebSockets/Http/WebSocketHandshaker.cs [hunk 1]
```diff
@@ -265,7 +267,7 @@ private void SendNegotiationErrorResponse(StreamWriter writer, HttpStatusCode co
             writer.Write("HTTP/1.1 ");
             writer.Write(intCode);
             writer.Write(" ");
-            writer.Write(HttpWorkerRequest.GetStatusDescription(intCode));
+            HttpStatusDescription.Get(code);
```
Comment (weweilep) [inline vtortola.WebSockets/Http/WebSocketHandshaker.cs]: I believe this should be wrapped in writer.Write().

Comment (StefH) [inline vtortola.WebSockets/Http/WebSocketHandshaker.cs]: Indeed. Good point. I'll fix asap.

**Comment by vtortola** [comment]
Any idea about how to make this work with AppVeyor?

```
Build started
git clone -q https://github.com/vtortola/WebSocketListener.git C:\projects\websocketlistener
git fetch -q origin +refs/pull/93/merge:
git checkout -qf FETCH_HEAD
nuget restore
MSBuild auto-detection: using msbuild version '14.0' from 'C:\Program Files (x86)\MSBuild\14.0\bin'.
TODO: unknown keyword platform
Command exited with code 1
```

**Comment by StefH** [comment]
I've some ideas.

I already do have another project which includes old .net and new .net core projects, you can take a look here:
https://github.com/StefH/XPath2.Net/blob/master/appveyor.yml

**Comment by StefH** [comment]
I've created an initial AppVeyor which builds 1 old .net and 1 new dotnet core project.

Also added a UnitTest which will automatically run when the code is build.

**Comment by vtortola** [comment]
Grand. I came back yesterday from hols, I will take a loot at this tomorrow and will merge if I see no problem. I am assuming this is taking latest code as base right? Cheers.

**Comment by StefH** [comment]
It's indeed based on your latests code.

If you also want to create new NuGet packages with that build batch file, maybe you need to update some versions in the project.json files.

But maybe just first merge the code, and then make it ready so new NuGets can be created.

**Comment by vtortola** [comment]
(I am stuck in Xmas preparations, I have this first thing in my list)

**Comment by vtortola** [comment]
@StefH I see you have replaced the unit tests, how come the former ones could not be reused? Cheers.

**Comment by StefH** [comment]
I just did not see them. Sorry. I'll add the existing ones the my unit test.

I keep you updated.

**Comment by StefH** [comment]
Added the Unit-Tests.

All pass, except one:
`public void With_WebSocketFrameHeaderFlags_Can_CreateMediumMaxHeader()`

I have the feeling that the test is wrong, the test verifies length 4, however the code generates a 10 length header?

**Comment by vtortola** [comment]
Back from Xmas. Thanks for that.

However there is something wrong... although it could be my machine configuration. It does not find any test to run.

```
C:\Users\vtortola\Documents\GitHub\WSL\WebSocketListener\tests\WebSocketListener.UnitTests>dotnet test
Project vtortola.Core.WebSocketListener (.NETStandard,Version=v1.3) was previously compiled. Skipping compilation.
Project WebSocketListener.UnitTests (.NETCoreApp,Version=v1.0) was previously compiled. Skipping compilation.
The specified framework 'Microsoft.NETCore.App', version '1.1.0' was not found.
  - Check application dependencies and target a framework version installed at:
      C:\Program Files\dotnet\shared\Microsoft.NETCore.App
  - The following versions are installed:
      1.0.1
  - Alternatively, install the framework version '1.1.0'.
SUMMARY: Total: 1 targets, Passed: 0, Failed: 1.
```

**Comment by StefH** [comment]
Do you have the latest dotnet version 1,1 SDK ?
https://www.microsoft.com/net/core#windowscmd

**Comment by vtortola** [comment]
Thanks. That was the problem 👍  (.NET core is something in my TODO list but I barely can find time ...).

I did a pull request to your branch fixing the problem with the test. Can you please confirm you see the issue solved? I will take a look tomorrow hopefully to see if there are any missing commit.

Thanks!

**Comment by StefH** [comment]
You PR to my branch fixes my issue with `With_WebSocketFrameHeaderFlags_Can_CreateMediumMaxHeader()`.

**Comment by deandob** [comment]
Hi, do post when these dotnet core changes are merged back in the main branch and available. Thanks for the time spent on this.

**Comment by weweilep** [comment]
Any update on this?  Would love to see a nuget package with .NET Standard support.

**Comment by vtortola** [comment]
Yes sorry, I have been very busy.

To be honest the problem is that I have no idea about .NET core and I am afraid of merging this and not being able of running my own code :D It is definitely something I want to get my hands dirty with but I have bigger battles to fight right now.

The code is published in StefH fork: https://github.com/StefH/WebSocketListener and can be used.

I will make some time to give it another go and make sure I know how to work with it in case something goes wrong.

**Comment by StefH** [comment]
Hi all,

I've uploaded my .NET Core NuGet packages to **MyGet** feed: please use that feed for testing this project.

https://www.myget.org/F/websocketlistener/api/v3/index.json

Issues or questions can be added to my github.

**Comment by StefH** [comment]
@weweilep did you have a chance to take a look at my NuGet which I published?

**Comment by weweilep** [comment]
@StefH I have, and it works fine with my .NET Standard 1.3 library.  I do get an error when trying to add it to a .NET 4.6 Framework application.

It adds the reference to the project, but it doesn't update the packages.config and gives the error: "Failed to add reference. The package 'vtortola.Core.WebSocketListener' tried to add a framework reference to 'System.Runtime' which was not found in the GAC. This is possibly a bug in the package. Please contact the package owners for assistance."

Despite failing, it still appears to work.

I should note, I'm not actually using this in a core environment, it is always being consumed/used on a .NET 4.6 Framework environment.

**Comment by StefH** [comment]
@weweilep Can you try again with latest 2.2.2.2 from https://www.myget.org/F/websocketlistener/api/v3/index.json ?

**Comment by weweilep** [comment]
@StefH I still get the same error of 'Failed to add reference. The package 'vtortola.Core.WebSocketListener' tried to add a framework reference to 'System.Runtime' which was not found in the GAC. This is possibly a bug in the package. Please contact the package owners for assistance.'

This can be replicated simply by creating a new .NET 4.6 console app and attempting to add the package through nuget.

**Comment by StefH** [comment]
Version 2,2.2.3 installs without any warning.

**Comment by weweilep** [comment]
Not sure if it's a verification or cache issue, but I show only 3 of the 4 packages updated to 2.2.2.3, the listener package has not yet updated to 2.2.2.3.

Those that are 2.2.2.3 are indeed adding successfully to .NET 4.5/4.6 applications.

**Comment by vtortola** [comment]
@StefH could you please help me to set up the testing for the core project and the build? I guess we have two builds, one for .NET and another one for .NET Core. The "WebSocketListener.UnitTests" should run only for the .NET build, and the "WebSocketListener.Core.UnitTests" should run only for the .NET Core.

**Comment by vtortola** [comment]
I will keep investigating why the some project appears on "tests" rather than in "Tests"

**Comment by StefH** [comment]
I did fix the projects, take a look a at my checkin.

About the core tests and normal tests:
I would suggest to use my app-veyor file (some fixes are needed probably), it should be possible to run both projects.
Maybe if that fails, keep my xunit tests for now, and we can investigate more if it's possible to run both.

**Comment by vtortola** [comment]
Thanks again for this 👍  We will take a look to the AppVeyor thing later.

**Comment by weweilep** [comment]
I see this was pulled to main, and I wasn't quite following how things were being named, but is it too late to suggest not using the name Core in the package and assembly names?  Technically, all of this was for .NETStandard, which .NET4.5/.NET Core/Mono/etc then implement (replacing PCL).

The nuspec files appear to have .Core on them which is going to cause upgrade issue as it's a different package name, and the old package name was fine.  Now that the new package has the proper lib folders specified, it doesn't seem needed to have .Core in the name.

Ideally, all assembly files should have the same name and be separated by their lib folder in the nuspec/nuget packages (which it does) and be one single package (for each type, WebSocketListener/Deflate/Rfc6455) and calling the existing library Core is rather confusing.

**Comment by StefH** [comment]
@vtortola 

1. I'll take a look at the appveyor.yml file, to get at least all code building again.
2. I did also restructure the project in a new [branch](https://github.com/StefH/WebSocketListener/tree/_stef_restructure), however I think it's indeed better that you do it by yourself to get familiar with all new things.
3. NuGet packages can be created with those `_build nuget.cmd` files

---

@weweilep 
This merge to main is not the final one. What needs to be done is "Make the new vs2015 "Core" projects the main projects. (Note that the new project will support .NETStandard and .NET4.5 and UAP. If possible we can also investigate if PCL is also possible.)

4. So rename the **Core** folder to **src** folder.
5. Rename all **.Core.**" folders (inside that **src** folder) to the normal name.
6. Move all code from the root projects "**vtortola.WebSockets.xxx**" to the "**src\vtortola.WebSockets.xxx**"
7. It's now possible to generate 4 NuGet packages.

---

I hope this clears some confusion.

**Comment by weweilep** [comment]
@StefH 
That clears it up perfectly, thank you.

The only issue I see is that the 4th package (WebSocketListener) doesn't really work as expected.  Instead of being no-code and bundling all project dependencies, it appears to pull the source for each project in and make a single dll with a different name instead of the 3 existing separate dlls (effectively an ILMerge).

In looking, I don't quickly see a way for 'dotnet pack' to make a non-code pack of dependencies which is unfortunate.

**Comment by StefH** [comment]
Indeed that 4th package (WebSocketListener) is just a new DLL which includes the code from all sub-projects. I actually like this approach also because : when is there a need to only include a single WebSocket.dll in your project ? Normally people would just include all 3.

But some investigation is still needed if it's possible to make a custom pack which includes indeed these 3 real dll's. (I already checked this, however I was not able to quickly fix this. Maybe in the new VS2017 this will be more easy)

**Comment by vtortola** [comment]
@StefH  there is some issue with the .Core.UnitTests project. If I unload it, everything works, but if I keep it loaded in the solution, it does not compile. It is currently breaking the build, and I have no idea what is wrong. Can you please take a look?

**Comment by StefH** [comment]
Hi Val.

What you can do is remove it temporary from the solution and appveyor file.

Btw : Is it breaking on your machine only?

I dont have time to investigate it now, sorry. After next week I can take a look.

**Comment by StefH** [comment]
@vtortola A question from me to you about github merging.
Maybe I did something wrong (or maybe not), but my fork is still indicating that I'm out of sync with your branch ? Like `This branch is 70 commits ahead, 3 commits behind vtortola:master.` Can you explain the 70 commits ahead ?

**Comment by vtortola** [comment]
Thanks for that!. I am not sure why, maybe because I did a squash merge?

**Pull request merged**